Fix split / rsplit collision

CHB-0r1s opened the pull request:
After research of all possible combinations, there is such schema for prevent useless mutations:
###  Without mutations
1) split() -> split()
2) rsplit() -> rsplit()
3) split(" ") -> split(" ")
4) rsplit(" ") -> rsplit(" ")
5) split(sep=" ") -> split(sep=" ")
6) rsplit(sep=" ") -> rsplit(sep=" ")
7) split(maxsplit=-1) -> split(maxsplit=-1)
8) rsplit(maxsplit=-1) -> rsplit(maxsplit=-1)
9) split(" ", maxsplit=-1) -> split(" ", maxsplit=-1)
10) rsplit(" ", maxsplit=-1) -> rsplit(" ", maxsplit=-1)

###  With mutations
1) split(maxsplit=1) -> rsplit(maxsplit=1)
2) rsplit(maxsplit=1) -> split(maxsplit=1)
3) split(" ", 1) -> rsplit(" ", 1)
4) rsplit(" ", 1) -> split(" ", 1)
5) split(" ", maxsplit=1) -> rsplit(" ", maxsplit=1)
6) rsplit(" ", maxsplit=1) -> split(" ", maxsplit=1)

I save it as a note in comments.

CHB-0r1s commented:
I hope this fix can help with https://github.com/boxed/mutmut/issues/389

CHB-0r1s commented:
@Otto-AA @boxed This magic with parsing -1 and +1 in libcst parser cause a lot of troubles for me. So, please, if you know better way to do it, I appreciate your help.

Otto-AA commented:
Hi, thank you for this PR!

> @Otto-AA @boxed This magic with parsing -1 and +1 in libcst parser cause a lot of troubles for me. So, please, if you know better way to do it, I appreciate your help.

I don't think that ` x.split(maxsplit=-1)` is written directly in the code. This is equivalent to ` x.split()`, which I would expect people to use, rather than explicitly providing and disabling `maxsplit`.

So I would suggest that we do not check if the value is `-1` and simply always mutate, if `maxsplit` is provided with any value. So `x.split(' ', -1)` would also be mutated to `x.rsplit(' ', -1)`, which is not optimal but should happen very seldomly.

Otto-AA commented on mutmut/node_mutation.py:
```diff
@@ -125,6 +128,62 @@ def operator_string_methods_swap(
+                if (
+                        (len(node.args) == 2 and "maxsplit" not in key_args) or
+                        ("maxsplit" in key_args and maxsplit_val != -1)
+                ):
```
We could rewrite it with the libcst matchers. I think both ways are quite readable, so choose whatever you prefer.

This would be the option with the libcst methods:

```python
                # either maxsplit is passed as the second arg (so 2+ args)
                if (m.matches(node, m.Call(args=[m.AtLeastN(n=2)]))
                    # or it is passed at any possition as a keyword arg
                    or m.matches(node, m.Call(args=[m.ZeroOrMore(), m.Arg(keyword=m.Name(value="maxsplit")), m.ZeroOrMore()]))):

                    func_name = cst.ensure_type(node.func, cst.Attribute).attr
                    yield node.with_deep_changes(func_name, value=new_call)
```

Otto-AA commented on mutmut/node_mutation.py:
```diff
@@ -125,6 +128,62 @@ def operator_string_methods_swap(
+                    continue
+
+                if (
+                        (len(node.args) == 2 and "maxsplit" not in key_args) or
```
It cannot happen that we have two args *and* `maxsplit` as a keyword arg (we cannot pass the argument both as a positional argument and keyword argument). So we could remove the ` and "maxsplit" not in key_args` part here.

Otto-AA commented on mutmut/node_mutation.py:
```diff
@@ -125,6 +128,62 @@ def operator_string_methods_swap(
+    """Try to handle specific mutations of string, which useful only in specific args combination."""
+
+    for old_call, new_call in supported_unsymmetrical_str_methods_swap:
+        if m.matches(node.func, m.Attribute(value=m.DoNotCare(), attr=m.Name(value=old_call))):
```
libcst does not verify the properties that we do not mention, so the `value=m.DoNotCare()` is superfluous.

```suggestion
        if m.matches(node.func, m.Attribute(attr=m.Name(value=old_call))):
```

CHB-0r1s commented:
Thank you @Otto-AA ! 
I will apply your suggestions as soon as possible today)

CHB-0r1s commented:
So, new schema

### Without mutations
1. split() -> split()
2. rsplit() -> rsplit()
3. split(" ") -> split(" ")
4. rsplit(" ") -> rsplit(" ")
5. split(sep=" ") -> split(sep=" ")
6. rsplit(sep=" ") -> rsplit(sep=" ")

### With mutations
1. split(" ", 1) -> rsplit(" ", 1)
2. rsplit(" ", 1) -> split(" ", 1)
3. split(maxsplit=1) -> rsplit(maxsplit=1)
4. rsplit(maxsplit=1) -> split(maxsplit=1)
5. split(maxsplit=-1) -> rsplit(maxsplit=-1)
6. rsplit(maxsplit=-1) -> split(maxsplit=-1)
7. split(" ", maxsplit=-1) -> rsplit(" ", maxsplit=-1)
8. rsplit(" ", maxsplit=-1) -> split(" ", maxsplit=-1)
9. split(" ", maxsplit=1) -> rsplit(" ", maxsplit=1)
10. rsplit(" ", maxsplit=1) -> split(" ", maxsplit=1)

CHB-0r1s commented:
Hello! I applied all suggestions. Code looks pretty easy now. Schema was removed from comments (add link).
Can you check new updates, please :3

@Otto-AA @boxed

CHB-0r1s commented:
Just little ping @Otto-AA
